src/v/iceberg/conversion/values_json.cc

```diff
@@ -0,0 +1,279 @@
+        }
+        ss::future<std::optional<value>> operator()(const struct_type& st) {
+            auto sv = std::make_unique<struct_value>();
+            co_await decode_struct(p, *sv, st, field_map);
+            co_return std::move(sv);
+        }
+        ss::future<std::optional<value>> operator()(const list_type& lt) {
+            auto lv = std::make_unique<list_value>();
+            co_await decode_list(p, *lv, lt, field_map);
+            co_return std::move(lv);
+        }
+        ss::future<std::optional<value>> operator()(const map_type&) {
+            throw value_conversion_exception(
+              "Map type is not supported in JSON deserialization");
+        }
+    };
+
+    v = co_await std::visit(
+      field_decoder_visitor{
+        p,
+        field_map,
+      },
+      ft);
+}
+
+ss::future<> decode_struct(
+  experimental::serde::json::parser& p,
+  struct_value& sv,
+  const struct_type& st,
+  const json_conversion_ir::struct_field_map_t& field_map) {
+    if (p.token() != experimental::serde::json::token::start_object) {
+        throw value_conversion_exception(fmt::format(
+          "Expected start of JSON object for struct type but got {}",
+          p.token()));
+    }
+
+    // Placeholder fields.
+    for (size_t i = 0; i < st.fields.size(); ++i) {
+        sv.fields.emplace_back();
+    }
+
+    while (true) {
+        if (!co_await p.next()) {
+            throw value_conversion_exception(
+              "Failed to read next JSON token after start object");
+        }
+
+        if (p.token() == experimental::serde::json::token::end_object) {
+            for (size_t i = 0; i < st.fields.size(); ++i) {
+                if (st.fields[i]->required && !sv.fields[i].has_value()) {
+                    throw value_conversion_exception(fmt::format(
+                      "Required field {} is missing", st.fields[i]->name));
+                }
+            }
+            co_return;
+        }
+
+        if (p.token() != experimental::serde::json::token::key) {
+            throw value_conversion_exception(fmt::format(
+              "Expected key token in JSON object but got {}", p.token()));
+        }
+
+        auto key = p.value_string();
+
+        auto key_len = key.size_bytes();
+        iobuf_parser iop(std::move(key));
+        auto key_as_string = iop.read_string(key_len);
+
+        auto it = field_map.find(key_as_string);
+
+        if (it == field_map.end()) {
+            // Key not found in the mapping, skip it.
+            // todo throw if extra keys not allowed?
+            // skip the value
+            co_await p.skip_value();
+            continue;
+        }
+
+        // Decode the value for this key.
+        if (!co_await p.next()) {
+            throw value_conversion_exception(
+              "Failed to read next JSON token after key");
+        }
+
+        co_await decode_field(
+          p,
+          sv.fields[it->second.field_pos],
```

Comment:
what prevents `it` from becoming invalid across `co_await` scheduling point? is it that the field_map remains unchanged for its lifetime?